### PR TITLE
Duplicate columns

### LINKED ISSUE #452: Extend column API to support duplicating columns
## Problem

The design spec for [Working with Columns](https://wiki.mathesar.org/en/design/specs/working-with-columns) calls for the ability to duplicate a column.

## Proposed solution

The Column creation endpoint should be extended to accept a reference to a another column in the same table. If this parameter is passed in, it should create a new column that's a copy of the column that has been passed in. All data in the column should also be copied.

## Additional context

- #324 
- Please ensure the parameters do not clash with #451 
- Mathesar's [API Standards](https://wiki.mathesar.org/en/engineering/architecture/api-standards)

### PR DESCRIPTION
**Pull request by eito-fis**
Fixes #452 

Adds the ability to duplicate columns on the backend, and exposes it via the API.

**Technical details**
Backend changes:
* Adds the `duplicate_column` function to `db.colums`. Allows us to create a duplicate of a column, with the choice of copying over data and or constraints from the source column. Note, if both data and constraints are not copied over, the new column will always be nullable.
* Adds the `get_column_constraints` function to `db.constraints`. Allows us to get the rows of `pg_constraint` for the passed column.
* Adds the `copy_constraint` function to `db.constraints`. Allows us to copy a constraint. Currently only supports unique constraints.

API changes:
* Extends the column creation endpoint with a set of new duplication parameters: `source_column`, `copy_source_data` and `copy_source_constraints`.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by eito-fis** [comment]
Adding duplication as an extension of the create endpoint was going to be pretty messy, so I ended us just adding a new endpoint. Let me know if this is fine, or if I we want to extend the create endpoint.

Also, it seems like the npm audit is failing again?

**Comment by kgodey** [comment]
@eito-fis Adding a new endpoint violates several of our [API standards](https://wiki.mathesar.org/en/engineering/architecture/api-standards)

> - A URL identifies a resource.
> - URLs should include nouns, not verbs.
> - Use plural nouns only for consistency (no singular nouns).
> - You shouldn’t need to go deeper than resource/identifier/resource.

Could you expand on why using the create endpoint would be messy?

I merged in `master`, which should fix the `npm audit`.

**Comment by eito-fis** [comment]
I've updated to extend the creation endpoint instead of having a separate duplication endpoint.

@kgodey  The validation gets a bit weird when we have two sets of parameters for the different operations, since we have to set all fields  to `required=False` and do manual validation.

File: mathesar/serializers.py
Comment (kgodey) [reply mathesar/serializers.py:77]: Also applies to the rest of the function.

File: mathesar/serializers.py
Comment (kgodey) [reply mathesar/serializers.py:89]: Also applies to the other error.

**Comment by kgodey** [comment]
I'd also like @mathemancer to review the `db` portions of things.

File: mathesar/serializers.py
Comment (eito-fis) [reply mathesar/serializers.py:73]: Sorry, not sure I understand this one. We need to check if both or neither set of arguments exist here, doesn't that mean we can't modularize?

Comment (kgodey) [reply mathesar/serializers.py:73]: @eito-fis I was thinking something like

```
def validate(self, data):
  self._validate_from_scratch_column(data)
  self._validate_duplicated_column(data)
  return data
```

I'm not sure if it would be worth it though, you might have too much repeated code.

File: mathesar/serializers.py
Comment (kgodey) [inline mathesar/serializers.py:102]: Please ensure this error matches the default DRF error for required fields not being passed in. I believe default DRF errors would return the field as the key and the error message as the value. See: https://www.django-rest-framework.org/api-guide/exceptions/

**Review by kgodey — APPROVED** [review]
@eito-fis Looks good to me except for one comment about validation errors. I'm approving to unblock merge, but please fix that before merge.

I haven't tested it locally, I'll leave that to @mathemancer since he'll also review this PR.

File: mathesar/serializers.py
Comment (eito-fis) [reply mathesar/serializers.py:102]: Given we have two distinct sets of required fields, what would this look like? Doing something like
```python
{
  'type': ['This field is required.'],
  'name': ['This field is required.'],
  'source_column': ['This field is required.'],
}
```
seems like it would be confusing (especially since name is optional for duplication). Would it make sense to have column from scratch as the "default" path, and return just
```python
{
  'type': ['This field is required.'],
  'name': ['This field is required.'],
}
```

Comment (kgodey) [reply mathesar/serializers.py:102]: You'd want to return an appropriate error based on what's passed in. e.g. if `copy_source_data` was passed in, you'd want to only return `'source_column': ['This field is required.'],`. if `type` was passed in, you'd want to return `'name': ['This field is required.'],`. You should be able to figure this out based on the required/specific fields you've already calculated.

Separately, we shouldn't require name for column creation, we could auto-generate one like we do for tables. If you don't want to handle that in this PR, please make a separate issue for it.

File: mathesar/serializers.py
Comment (eito-fis) [reply mathesar/serializers.py:73]: I think my confusion is more from that it seems like our conditions are dependent on both `from_scratch` and `duplicated` and therefore seems hard to separate? Ex: the first condition ensures there are only fields specific to a single path, which requires looking at the arguments for both paths at the same time.

Comment (kgodey) [reply mathesar/serializers.py:73]: Yeah, that's why I think there would be too much repeated code between the two and it might not be worth it. Feel free to leave it as-is.

File: mathesar/serializers.py
Comment (eito-fis) [reply mathesar/serializers.py]: Updated the error, with it defaulting to from scratch if no parameters were passed. Also opened #555.

**Comment by kgodey** [comment]
Since Brent is out, I'm merging this. I did test locally and it looks good 👍

**Pull request merged**
